### ISSUE TITLE
Mock classification submission systematics

### ISSUE THREAD

**Issue by aimalz** [aimalz/proclam#4]
_Simulate some mock classification results (with a goal of exaggerating intuitively good or bad results to check if metrics successfully reflect those)._

EDIT: That was a very general goal!  This abandoned issue is being transformed into a hub for covering all the cases of mock classification results that we want to check.  

The systematics we want to consider are:
- [ ]  Perfect classifier: returns 1 at the true class and 0 elsewhere, as a "control" that should beat every other classifier
- [x] Uniform guesser: takes a uniform random guess at the probability of each type and then normalizes it, as a "control" that should be very easy to beat in the absence of pathological systematics
- [x] Tunnel vision classification: performs well on only one class and ignores the others
- [x] Broadbrush classification: performs well on the most populous class(es) but poorly on less common classes
- [x] Cruise control classification: overwhelmingly predicts that all objects are of one class
- [ ] Covariant classification: consistently misclassifies a small number of classes with one another but effectively distinguishes them from other classes
- [x] Subsumed classification: consistently misclassifies class m as class m'

Please comment with any other cases of whose behavior under a metric should be investigated.  

EDIT: __To make it easy for others to quickly glance through the issue and see what other cases have been considered, please end a comment suggesting a new classifier with a markdown checkbox and shorthand name and description, and then check off the box when you've finished making its module.__

EDIT: I'm splitting up the systematics of the classifier from its format -- for the latter, see #14.

**Comment by aimalz** [comment]
~~I'll implement a vanilla version with uncorrelated, low-scatter predictions.~~  This ended up being a very stupid classifier that ignores the data and takes a totally random guess at the probability of each type from a uniform distribution (and then normalizes it to sum to 1).

**Comment by aimalz** [comment]
We should also implement a perfect classifier as a control for our tests.

**Comment by aimalz** [comment]
As @gnarayan brought up in today's call, there should be a classifier that takes as input a confusion matrix as well as the truth table.

EDIT: I split this into #19 to distinguish between the pathologies to investigate and the infrastructure to support them.  #19 can enable many of the items in the list for this issue.

**Comment by Andromedanita** [comment]
@aimalz I started looking into the codes for classifying by guessing and using confusion matrix. I want to implement tunnel and broadbrush  classifiers but I have two questions:

1- For tunnel classifier, we want the classifier to classify only one class correctly and ignore other classes.  Can it classify other classes correctly by chance or do we want it to classify those classes incorrectly all the time?

2- Would it work by just taking one of the classes randomly (in tunnel classifier case) or taking the most common class (in broadbrush classifier) and assigning the probability for that class to be a random number larger than 1/(# of classes) ?

**Comment by aimalz** [comment]
@Andromedanita Good questions!

What I was imagining for the tunnel vision classifier was that it would classify other classes agnostically, so occasionally getting it right by random chance.  I think we'd also want to test this for a range of accuracies on the focus class.

For both of them, I was thinking of using the `from_cm` classifier.  For the tunnel vision classifier, the confusion matrix would have one column that's close to one on the diagonal with low numbers on the off-diagonal, and the other columns would be filled with values close to 1/M for M classes.  For the broadbrush classifier, it might have a scaling like a diagonal component that's proportional to M, add some small random scatter to perturb it from perfection, and then normalize it so the rows sum to 1 (I think).

Does that make sense?

**Comment by Andromedanita** [comment]
Thanks @aimalz , it makes sense I think. Just to clarify, we want the rows to sum to 1 in the tunnel vision classifier as well, correct? Because those are the probabilities for the data point to be in each class.

**Comment by aimalz** [comment]
@Andromedanita That is correct.  The `from_cm` classifier assumes the input confusion matrix is already normalized such that the rows each sum to 1.

**Comment by aimalz** [comment]
I added another case for a "cruise control" classifier that's sort of a combination of broadbrush and tunnel vision classifiers, but could be a realistic case for the kind of classifier that only cares about one type of object and instead of giving equal probability to all other classes when it thinks a given object is not of its favorite class, it would instead more or less assign them all to one "junk" class.  If we weighted equally per object, a metric would favor this classifier in the case that its favorite class was the most common, but it should take a hit if we equally weight per class.  Thanks @Andromedanita for the idea!
